statsmodels/tsa/exponential_smoothing/ets.py

```diff
@@ -0,0 +1,1288 @@
+            param_names += ['initial_level']
+            if self.has_trend:
+                param_names += ['initial_trend']
+            if self.has_seasonal:
+                param_names += [
+                    f'initial_seasonal.{i}'
+                    for i in range(self.seasonal_periods)
+                ]
+        return param_names
+
+    @property
+    def _internal_param_names(self):
+        param_names = [
+            'smoothing_level',
+            'smoothing_trend',
+            'smoothing_seasonal',
+            'damping_trend',
+            'initial_level',
+            'initial_trend',
+        ]
+        param_names += [
+            f'initial_seasonal.{i}' for i in range(self.seasonal_periods)
+        ]
+        return param_names
+
+    @property
+    def _k_states(self):
+        return (
+            1  # level
+            + int(self.has_trend)
+            + int(self.has_seasonal)
+        )
+
+    @property
+    def _k_smoothing_params(self):
+        return self._k_states + int(self.damped_trend)
+
+    @property
+    def _k_initial_states(self):
+        return (
+            1 + int(self.has_trend) +
+            + int(self.has_seasonal) * self.seasonal_periods
+        )
+
+    @property
+    def k_params(self):
+        k_params = self._k_smoothing_params
+        if self.initialization_method == 'estimated':
+            k_params += self._k_initial_states
+        return k_params
+
+    @property
+    def _k_params_internal(self):
+        return 4 + 2 + self.seasonal_periods
+
+    def _internal_params(self, params):
+        """
+        Converts a parameter array passed from outside to the internally used
+        full parameter array.
+        """
+        # internal params that are not needed are all set to zero, except phi,
+        # which is one
+        internal = np.zeros(self._k_params_internal)
+        for i, name in enumerate(self.param_names):
+            internal_idx = self._internal_params_index[name]
+            internal[internal_idx] = params[i]
+        if not self.damped_trend:
+            internal[3] = 1  # phi is 4th parameter
+        return internal
+
+    def _model_params(self, internal):
+        """
+        Converts internal parameters to model parameters
+        """
+        params = np.empty(self.k_params)
+        for i, name in enumerate(self.param_names):
+            internal_idx = self._internal_params_index[name]
+            params[i] = internal[internal_idx]
+        return params
+
+    def _set_fixed_params(self, params):
+        if self._has_fixed_params:
+            for i, name in enumerate(self._fixed_params):
+                idx = self._fixed_params_index[i]
+                params[idx] = self._fixed_params[name]
+
+    @property
+    def _start_params(self):
+        lb = []
+        ub = []
+        for b in self.bounds:
+            lb.append(b[0])
+            ub.append(b[1])
+
+        # See https://github.com/robjhyndman/forecast/blob/master/R/ets.R
```

Comment:
Can you provide a different reference? forecast it GPL licensed and so code from it cannot be used or directly translated.

Reply:
If this appears in any written document that isn't code, it would be sufficient.

Reply:
I removed this reference and am now using the start params similar to the implementation in the state space module.
This was more of an experiment, I was unaware that GPL also forbids translations.

Reply:
It doesn't forbid them, but direct translations should result in code that is also GPL'd. The PyData econsystem uses BSD/MIT and not GPL.